### DISCUSSION TITLE
How to fix "ImproperlyConfiguredException: 'args' does not have a type annotation"?

### DISCUSSION THREAD

**Discussion by danielniccoli** [litestar-org/litestar#3014] [Q&A]
I try to write an Azure Durable Function based on LiteStar. I'm using the Azure Durable Functions SDK for that. To use the Durable Functions client, you add a decorator to a function and set the client name. Through the decorator, the client is then passed as an argument to the function. I don't understand the details, but this is how that looks like:

```py
@get("/approve/{id:str}")
# The decorator below is necessary and passes the Orchestration Client to the function.
# You also need to add `df_client` as an argument to the function.
# However, doing so causes the application to raise an Exception:
#     ImproperlyConfiguredException: 500: 'args' does not have a type annotation. If it should receive any value, use 'Any'.
@app.durable_client_input(client_name="df_client")
async def approve(id: str, client: df.DurableOrchestrationClient) -> str:
    # Do something with df_client here, like starting a new orchestration.
    return id

litestar_app.register(approve)
```

The problem is, once I add the decorator, the following exception is raised:

```
Exception: ImproperlyConfiguredException: 500: 'args' does not have a type annotation. If it should receive any value, use 'Any'.
```

I don't exactly know why, but I assume it is because the decorator changes the function so it has arguments without proper type hinting. How do I solve this without changing any code on the Durable Functions SDK?

**Comment by Alc-Alc** [comment]
This might be a naive question to a regular user of "Azure Durable Functions", but I checked https://learn.microsoft.com/en-us/azure/azure-functions/durable/quickstart-python-vscode?tabs=macos%2Cazure-cli-set-indexing-flag&pivots=python-mode-decorators and I fail to see anything related to a web app. Can you clarify to me (and maybe others) how an asgi app comes into play with Durable functions?

On a side note, I did a cursory search and came across https://learn.microsoft.com/en-us/samples/azure-samples/fastapi-on-azure-functions/azure-functions-python-create-fastapi-app/ which seems to work for asgi apps in general, but I don't know if that is what you want

**Reply by danielniccoli** [reply]
The asgi app should process the http request and call the durable client. The FastAPI example is exactly the way how you would use FastAPI or Litestar for that matter. But that's not the problem I have. The problem is that Litestar raises an `ImproperlyConfiguredException` for the reasons I described above and I need a way to fix that.

**Reply by provinzkraut** [reply]
> The asgi app should process the http request and call the durable client. The FastAPI example is exactly the way how you would use FastAPI or Litestar for that matter. But that's not the problem I have. The problem is that Litestar raises an ImproperlyConfiguredException for the reasons I described above and I need a way to fix that.

I think it might actually be the problem you have. I'm not an expert on azure functions, but after browsing through the docs, I'm not sure you're supposed to handle ASGI apps like that. You're kinda mixing both layers (azure function app ans ASGI app) here. From what I understand, the azure function app is responsible for calling into the ASGI app, which means you should handle the non-Litestar stuff in an underlying functions app before you delete. What you are doing here is asking Litestar to handle azure functions related things, which it (obviously) doesn't know how to do.

I found a similar discussion to your problem here: https://github.com/Azure/azure-functions-durable-python/issues/444

As for the issue, your assessment sounds logical to me. `durable_client_input` is not a well behaved decorator and obscures the function it decorates, but Litestar relies on the signature information from that function. There is nothing we can do here to fix it, since this problem does not originate in Litestar.

If you need further assistance with this, an MCVE or a more complete example would help to get a picture of how you've set up things.

**Reply by danielniccoli** [reply]
> There is nothing we can do here to fix it, since this problem does not originate in Litestar.

But maybe it can be solved in Litestar? Is there a way I can tell litestar manually the signature of that function?
